### PR TITLE
adds virtual scroll adapter

### PR DESCRIPTION
**Pull request by ben-henoida**
As discussed in  #10122 

I order to integrate virtual scroll into mat table we need to decouple CdkVirtualScrollViewport from  CdkVirtualForOf

This modification introduces an adapter between the 2 classes, so users can provide their own version of 
CdkVirtualScrollAdapter

suggested by @shlomiassaf,

### CODE REVIEW

**Comment by ben-henoida** [comment]
I signed it!

**Comment by IlCallo** [comment]
Related to this attack of some days ago:
https://github.com/dominictarr/event-stream/issues/116

That package was malicious code and has been removed from the registry

**Comment by DevVersion** [comment]
We've fixed this issue with #14295. If you rebase your PR, the tests should run as you would expect.

**Comment by ben-henoida** [comment]
Thank you @IlCallo for the info.
Do you know if the bot will retry the checks once this is fixed ?

**Comment by IlCallo** [comment]
As dev said, you should fetch from upstream, rebase and force push.

Some useful instructions are written into CONTRIBUTING file, I think

**Comment by mmalerba** [comment]
Looks good, but I would like if we could come up with a better name than `CdkVirtualScrollAdapter`. @jelbourn @crisbeto any naming ideas?

File: src/cdk/scrolling/public-api.ts [hunk 1]
```diff
@@ -14,3 +14,4 @@ export * from './viewport-ruler';
 export * from './virtual-for-of';
 export * from './virtual-scroll-strategy';
 export * from './virtual-scroll-viewport';
+export * from './virtual-scroll-adapter';
```
Comment (crisbeto) [inline src/cdk/scrolling/public-api.ts]: Needs an extra newline at the end of the file.

File: src/cdk/scrolling/virtual-scroll-adapter.ts [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {Observable} from 'rxjs';
+import {ListRange} from '../collections';
+
+export interface CdkVirtualScrollAdapter<T> {
```
Comment (crisbeto) [inline src/cdk/scrolling/virtual-scroll-adapter.ts]: This needs some docs explaining what it does and what it's useful for.

File: src/cdk/scrolling/virtual-scroll-adapter.ts [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+  dataStream: Observable<T[] | ReadonlyArray<T>>;
```
Comment (crisbeto) [inline src/cdk/scrolling/virtual-scroll-adapter.ts]: @mmalerba shouldn't we call this a `dataSource` for consistency with some of the other components?

Comment (mmalerba) [reply src/cdk/scrolling/virtual-scroll-adapter.ts]: Actually this is different than the `dataSource`, in the case of `cdkVirtualForOf` this stream _is_ created by connecting to the data source, but it doesn't have to be.

File: src/cdk/scrolling/virtual-scroll-adapter.ts (hunk 2 above)
Comment (jelbourn) [inline src/cdk/scrolling/virtual-scroll-adapter.ts]: How about `CdkVirtualScrollRepeater`?

**Comment by IlCallo** [comment]
Any advancement on this?

**Comment by CRACKISH** [comment]
How are you doing on this request? I need this feature ))

**Comment by CharlBest** [comment]
Will this pull request help with https://github.com/angular/material2/issues/13862 and if so is there any timeline to get this in?

**Comment by Splaktar** [comment]
@ben-henoida thank you for making your first contribution with this PR. Can you please address some of the review feedback above? Please let us know if you run into any issues or have questions about how to do that.

**Comment by mmalerba** [comment]
Closing due to inactivity

**Pull request closed without merging**